backend/src/layer/logic/api_command/auth/authenticator.rs

```diff
@@ -0,0 +1,134 @@
+use base64::{engine::general_purpose::STANDARD, Engine};
+
+use sha2::{Digest, Sha512};
+
+use crate::{
+    interface::api_command::{CommandError, InnerAuthInfo, Result},
+    util::Uuid,
+};
+
+use super::image_command_type::ImageCommandType;
+
+/// Class for authenticating commands.
+pub struct Authenticator {
+    api_keys: Vec<String>,
+}
+
+impl Authenticator {
+    /// Creates a new object which authenticates requests against the given list of valid api keys.
+    #[must_use]
+    pub fn new(api_keys: Vec<String>) -> Self {
+        Self { api_keys }
+    }
+
+    /// Authenticates image commands by checking if the given hash conforms with the given api key.
+    /// # Errors
+    /// [`CommandError::BadAuth`] is returned when authentication failed.
+    pub fn authn_image_command(
+        &self,
+        auth_info: &InnerAuthInfo,
+        image_id: Uuid,
+        image_command_type: ImageCommandType,
+    ) -> Result<()> {
+        let report_reason = if let ImageCommandType::ReportImage(reason) = image_command_type {
+            reason.to_string()
+        } else {
+            String::new()
+        };
+
+        let hash = self.calculate_hash(
+            auth_info,
+            &image_command_type.to_string(),
+            &[&image_id, &report_reason],
+        )?;
```

Comment:
Calculate hash in external function. `gen_hash_image_cmd(...) -> shall`. Thinks like that could be repetitive and usually be determinded isolated. For changes in generation, it's also better to change the functions as each part in the auth functions.